Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm

```diff
@@ -103,20 +118,42 @@ - (void)displayLinkFired:(CADisplayLink *)sender
     _drawingAreaProxy->didRefreshDisplay();
 }
 
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+- (void)timerFired
+{
+    ASSERT(isUIThread());
+    if (_updateTimerPaused || ![_updateTimer isValid])
+        return;
+
+    _drawingAreaProxy->didRefreshDisplay();
+}
+#endif // ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+
 - (void)invalidate
 {
     [_displayLink invalidate];
     _displayLink = nullptr;
+
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+    [_updateTimer invalidate];
+    _updateTimer = nil;
+#endif
 }
 
 - (void)schedule
 {
     _displayLink.paused = NO;
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+    _updateTimerPaused = NO;
```

Comment:
Rather than this, why not just destroy the timer in `pause`, and recreate it here?

Reply:
I wasn’t sure if there was a 1:1 mapping of `pause` to `schedule` calls but I could invalidate the timer which should remove it from the run loop in `pause` and then re-create the timer here if `![_updateTimer isValid]`

Reply:
That would require either caching the update rate or reading the NSUserDefault again, the latter should be fine as I don’t think the call is frequent